### PR TITLE
[GStreamer][MSE] Avoid seeking in READY or NULL pipeline state

### PR DESCRIPTION
**Pull request by ybandou**
#### 2936a3b532ac227f0f996e7032b3f9fb130d2b96
<pre>
[GStreamer][MSE] Avoid seeking in READY or NULL pipeline state
<a href="https://bugs.webkit.org/show_bug.cgi?id=245358">https://bugs.webkit.org/show_bug.cgi?id=245358</a>

Reviewed by NOBODY (OOPS!).

We should put the seek request on pending when the GStreamer pipeline is in a state below PAUSED state.

Before the PAUSED state, the hardware devices are not ready to doing a seek and reset the devices,
such as the video decoder, knowing that in the READY state we reserve and open the devices asynchronously.

We also need to avoid performing a new seek when another is pending, this behavior improves robustness when
performing many searches in a short period of time.

* Source/WebCore/platform/graphics/gstreamer/mse/MediaPlayerPrivateGStreamerMSE.cpp:
(WebCore::MediaPlayerPrivateGStreamerMSE::seek):
(WebCore::MediaPlayerPrivateGStreamerMSE::doSeek):
(WebCore::MediaPlayerPrivateGStreamerMSE::asyncStateChangeDone):
(WebCore::MediaPlayerPrivateGStreamerMSE::updateStates):
</pre>
https://github.com/WebKit/WebKit/commit/7c8fbfa7e7df91543b5ea27c23f4f48d1706948e

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/218 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/26/builds/230 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/14/builds/240 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/218 "Built successfully") | [✅ 🛠 wincairo](https://ews-build.webkit.org/#/builders/32/builds/197 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/219 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/23/builds/350 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/16/builds/556 "Built successfully") | [  ~~🧪 wpe-wk2~~](https://ews-build.webkit.org/#/builders/34/builds/464 "Build was cancelled. Recent messages:Failed to print configuration; Skipping applying patch since patch_id isn't provided; Checked out pull request; Skipped layout-tests; layout-tests (exception)") | 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/223 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/15/builds/288 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/219 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/239 "Built successfully") | 
| | [❌ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/310 "2 api tests failed or timed out") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/10/builds/215 "Passed tests") | [  ~~🧪 gtk-wk2~~](https://ews-build.webkit.org/#/builders/1/builds/291 "Build was cancelled. Recent messages:Printed configuration; Skipping applying patch since patch_id isn't provided; Checked out pull request") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/7/builds/186 "Built successfully") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/36/builds/226 "Passed tests") | [⏳ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/API-Tests-GTK-EWS "Waiting to run tests") | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/30/builds/208 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/17/builds/453 "Passed tests") | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/4/builds/179 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/221 "Passed tests") | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/31/builds/211 "Built successfully") | | | 
| | | | |

### CODE REVIEW

**Comment by philn** [comment]
> Before the PAUSED state, the hardware devices are not ready to doing a seek and reset the devices,
such as the video decoder, knowing that in the READY state we reserve and open the devices asynchronously.

Well, what you describe is for some specific GStreamer elements I suppose? That behavior is not standard. Devices should be opened during the NULL to READY transition. Not after.

**Comment by ybandou** [comment]
From GStreamer's point of view, is it normal to doing a seek in READY state ? without receiving the content metadata.

**Comment by philn** [comment]
> From GStreamer's point of view, is it normal to doing a seek in READY state ? without receiving the content metadata.

Quoting the gst_element_seek_simple docs,

In a completely prerolled PAUSED or PLAYING pipeline, seeking is always guaranteed to return [TRUE](https://developer.gnome.org/glib/unstable/glib-Standard-Macros.html#TRUE:CAPS) on a seekable media type or [FALSE](https://developer.gnome.org/glib/unstable/glib-Standard-Macros.html#FALSE:CAPS) when the media type is certainly not seekable (such as a live stream).

Some elements allow for seeking in the READY state, in this case they will store the seek event and execute it when they are put to PAUSED. If the element supports seek in READY, it will always return [TRUE](https://developer.gnome.org/glib/unstable/glib-Standard-Macros.html#TRUE:CAPS) when it receives the event in the READY state.

**Comment by ybandou** [comment]
I agree, but currently, in MSE implementation, we send the seek event directly to the WebKitMediaSrc element. 
Consenquetly the elements don't receive the seek event and receive directly the FLUSH event, thus they can't store the seek event when they are not ready.

**Comment by philn** [comment]
CCing @eocanha our MSE/seek expert ;)

**Comment by eocanha** [comment]
After [Alicia's refactoring](https://github.com/WebKit/WebKit/commit/9382c6ef1a5b64d1701c691c63fe84de22cf932b), she's the one completely aware of all the implications of seeks now, but she won't be able to review this at least until next week.

I see that Alicia's refactoring has completely removed the usage of m_isSeekPending. What I've seen from this PR looks fine, but I can't be sure if I'm missing any side effect. Still, I have some questions for specific lines of code.

File: Source/WebCore/platform/graphics/gstreamer/mse/MediaPlayerPrivateGStreamerMSE.cpp [hunk 1]
```diff
@@ -159,7 +159,34 @@ MediaTime MediaPlayerPrivateGStreamerMSE::durationMediaTime() const
 void MediaPlayerPrivateGStreamerMSE::seek(const MediaTime& time)
 {
     GST_DEBUG_OBJECT(pipeline(), "Requested seek to %s", time.toString().utf8().data());
-    doSeek(time, m_playbackRate, GST_SEEK_FLAG_FLUSH);
+
+    if (m_isSeekPending) {
+        if (m_seekTime != time) {
+            m_seekTime = time;
+            m_mediaSource->seekToTime(m_seekTime);
```
Comment (eocanha) [inline Source/WebCore/platform/graphics/gstreamer/mse/MediaPlayerPrivateGStreamerMSE.cpp:234]: Why do you need to call seekToTime() here...

File: Source/WebCore/platform/graphics/gstreamer/mse/MediaPlayerPrivateGStreamerMSE.cpp [hunk 2]
```diff
@@ -159,7 +159,34 @@ MediaTime MediaPlayerPrivateGStreamerMSE::durationMediaTime() const
...
+    if (getStateResult == GST_STATE_CHANGE_FAILURE || getStateResult == GST_STATE_CHANGE_NO_PREROLL) {
+        GST_DEBUG_OBJECT(pipeline(), "[Seek] cannot seek, current state change is %s", gst_element_state_change_return_get_name(getStateResult));
+        return;
+    }
+    if (state < GST_STATE_PAUSED) {
+        m_isSeekPending = true;
+        m_isSeeking = true;
+        m_seekTime = time;
+        m_isEndReached = false;
+        m_mediaSource->seekToTime(m_seekTime);
```
Comment (eocanha) [inline Source/WebCore/platform/graphics/gstreamer/mse/MediaPlayerPrivateGStreamerMSE.cpp:250]: ...and here...

File: Source/WebCore/platform/graphics/gstreamer/mse/MediaPlayerPrivateGStreamerMSE.cpp [hunk 3]
```diff
@@ -183,13 +211,13 @@ bool MediaPlayerPrivateGStreamerMSE::doSeek(const MediaTime& position, float rat
 
     // Important: In order to ensure correct propagation whether pre-roll has happened or not, we send the seek directly
     // to the source element, rather than letting playbin do the routing.
-    gst_element_seek(m_source.get(), rate, GST_FORMAT_TIME, seekFlags,
+    ret = gst_element_seek(m_source.get(), rate, GST_FORMAT_TIME, seekFlags,
         GST_SEEK_TYPE_SET, toGstClockTime(m_seekTime), GST_SEEK_TYPE_NONE, 0);
     invalidateCachedPosition();
 
     // Notify MediaSource and have new frames enqueued (when they're available).
     m_mediaSource->seekToTime(m_seekTime);
```
Comment (eocanha) [inline Source/WebCore/platform/graphics/gstreamer/mse/MediaPlayerPrivateGStreamerMSE.cpp]: ^^^ This is line 219

File: Source/WebCore/platform/graphics/gstreamer/mse/MediaPlayerPrivateGStreamerMSE.cpp [hunk 4]
```diff
@@ -285,6 +314,16 @@ void MediaPlayerPrivateGStreamerMSE::updateStates()
             GST_ERROR_OBJECT(pipeline(), "Setting the pipeline to PAUSED failed");
         m_isPipelinePlaying = false;
     }
+
+    if (m_isSeekPending) {
+        GST_DEBUG_OBJECT(pipeline(), "[Seek] committing pending seek to %s", toString(m_seekTime).utf8().data());
+        m_isSeekPending = false;
+        m_isSeeking = doSeek(m_seekTime, m_player->rate(), static_cast<GstSeekFlags>(GST_SEEK_FLAG_FLUSH));
```
Comment (eocanha) [inline Source/WebCore/platform/graphics/gstreamer/mse/MediaPlayerPrivateGStreamerMSE.cpp:400]: ...if doSeek() is going to happen eventually and call seekToTime() eventually in line 219?

**Review by eocanha — COMMENTED** [review]
After [Alicia's refactoring](https://github.com/WebKit/WebKit/commit/9382c6ef1a5b64d1701c691c63fe84de22cf932b), she's the one completely aware of all the implications of seeks now, but she won't be able to review this at least until next week.

I see that Alicia's (@ntrrgc) refactoring has completely removed the usage of m_isSeekPending. What I've seen from this PR looks fine, but I can't be sure if I'm missing any side effect. Still, I have some questions for specific lines of code.

File: Source/WebCore/platform/graphics/gstreamer/mse/MediaPlayerPrivateGStreamerMSE.cpp (hunk 1 above)
Comment (ybandou) [reply Source/WebCore/platform/graphics/gstreamer/mse/MediaPlayerPrivateGStreamerMSE.cpp:234]: In order to anticipate the content download from the seek position in the MSE part.
This PR put just the seek request on pending in the GStreamer part and not in the MSE part.

File: Source/WebCore/platform/graphics/gstreamer/mse/MediaPlayerPrivateGStreamerMSE.cpp (hunk 4 above)
Comment (ybandou) [reply Source/WebCore/platform/graphics/gstreamer/mse/MediaPlayerPrivateGStreamerMSE.cpp:400]: Yes I agree if the seekToTime() is already called, we won't call it again in doSeek().
I will change this and i will test it.

**Comment by blino** [comment]
For reference, this relates to the following downstream MR/issue:
- https://github.com/WebPlatformForEmbedded/WPEWebKit/issues/867
- https://github.com/WebPlatformForEmbedded/WPEWebKit/pull/868

File: Source/WebCore/platform/graphics/gstreamer/mse/MediaPlayerPrivateGStreamerMSE.cpp (hunk 4 above)
Comment (ybandou) [reply Source/WebCore/platform/graphics/gstreamer/mse/MediaPlayerPrivateGStreamerMSE.cpp:400]: Is done.

**Comment by eocanha** [comment]
I'm not a reviewer. Pinging @ntrrgc, who is.

**Comment by ybandou** [comment]
@ntrrgc @philn  Could you please see this pull request ?

**Review by ntrrgc — CHANGES_REQUESTED** [review]
This patch should not be necessary with the current WebKitMediaSrc upstream: handling initial seeks was a design goal of the current (reworked) WebKitMediaSrc.

An initial seek doesn't cause a seek event to be propagated downstream. Instead, the initial segment sent by WebKitMediaSrc will have the time adjusted to start at the requested position.

The idea here is: if a user requested playback to start at a certain position before even any media data was available, it makes sense to start playback at that position, rather than starting playback at a completely different position (0) and trying to change it after the fact.

These are the steps involved when an initial seek occurs with the current code:

1. `MediaPlayerPrivateGStreamerMSE::seek(time)` is called, which calls `doSeek()` and in turn request a seek to the WebKitMediaSrc element by using `gst_element_seek()`.
2. That GStreamer call in turn calls `webKitMediaSrcSendEvent()` synchronously (from the same main thread), which calls `webKitMediaSrcSeek()`. (Note that since it's only WebKit requesting seeks, relying on the GStreamer event API for this was actually optional. I could have exposed `webKitMediaSrcSeek()` directly as part of the public API of WebKitMediaSourceGStreamer.h and the result would be the same.)
3. `webKitMediaSrcSeek()` updates `source->priv->startTime` with the requested time. Note this field is only read and written from the main thread. Then, it calls `webKitMediaSrcStreamFlush(isSeekingFlush=true)` for each track.
4. `webKitMediaSrcStreamFlush()` takes the streaming members DataMutex and checks if the first object (~sample) has been popped by the WebKitMediaSrc loop function. If it has, then this is a normal seek that will cause a FLUSH event to be propagated downstream. If it hasn't then we have the following guarantees:
    1. No SEGMENT event has been pushed downstream by the WebKitMediaSrc pad yet. This is guaranteed because a SEGMENT event is only sent after the first object has been popped from the queue -- but before the object itself is pushed downstream. In other words, the loop function intentionally delays sending a SEGMENT event until we have any frames, so that we don't have to send unnecessary segments that could expose bugs in elements downstream.
    2. No new object will be pushed into the queue while we are in this function, as objects are pushed from the main thread, which is now busy seeking.
    3. The WebKitMediaSrc pad loop function is not making use of the streaming members, as we just took posession of the DataMutex. In most cases, it will be blocked waiting for `streamingMembers->queueChangedOrFlushedCondition`.
5. Assuming the case of an initial flush, the queue is cleared and we set the local variable `skipFlush = true`. After that, the stream DataMutex is released temporarily, but note that  the streaming thread won't be pushing any frames or events just yet, as the queue is now guaranteed to be empty and we still have the main thread busy with the seek.
6. The `skipFlush` flag makes us skip the `if()` blocks where the flush events would be sent downstream, but we still run the code that takes the DataMutex again and updates the track segment to use the new `startTime`.
7. We return from everything. The initial seek is done. The pad loop function didn't have to do anything in the whole process, and in fact chances are it wasn't even awaken.

This is what will happen later when an append is done:

1. When an append is done containing currentTime, the multiplatform code calls `SourceBufferPrivateGStreamer::enqueueSample()`, which in turn enqueues the sample into the TrackQueue (see `MediaSourceTrackGStreamer::enqueueObject()` -> `TrackQueue::enqueueObject()`.
2. `TrackQueue::enqueueObject()` will invoke the `notEmptyCallback` directly *from the main thread*, if it's set. This gives us two possible code paths indee:
  1. There is a `notEmptyCallback` set. In this case the streaming thread is blocked waiting on the `streamingMembers->queueChangedOrFlushedCondition`:
      1. The callback is a lambda function inside `webKitMediaSrcLoop()` which takes the `streamingMembersDataMutex`, sets `hasPoppedFirstObject` flag to true and writes the sample to the captured variable `object` which the streaming thread can read. Then it notifies `queueChangedOrFlushedCondition`, which awakens the streaming thread.
      2. The streaming thread is awoken with posession of the streaming members mutex and checks the condition variable's condition: `streamingMembers->isFlushing || object`. The check is true because `object` is not null anymore.
  3. There isn't a `notEmptyCallback` set. In this case the streaming thread hasn't checked the TrackQueue yet, but when it does, it will get the object without needing to wait.
3. No matter what codepath we needed to take, both will converge on the streaming thread having the streaming members DataMutex taken and the first sample as `object`.
4. `doesNeedSegmentEvent` is true as no segment has been sent. The segment is sent downstream now, while intentionally keeping streaming members locked to disallow flushes while it's being propagated downstream. This is done with the assumption that the segment propagation should not block, and we would rather risk not hitting race conditions in downstream elements arising from a flush mid first segment. Then `doesNeedSegmentEvent` is set to false.
5. Caps are checked against the previous caps. If they are the same (initial caps have already been sent, obtained from the MSE initialization segments), nothing is done. Otherwise, we send a `caps` event downstream with the streaming members DataMutex taken, simply because we don't want a flush mid-caps if we can avoid it for downstream elements stability reasons.
6. The DataMutex is finally released and the sample (in the variable `object`) is finally chained downstream.

Woah, that took much longer to explain than I hoped for! But after all **seeks and flushes are very tricky.** I spent a lot of effort carefully analyzing all the potential execution paths I could think when working on this reworked WebKitMediaSrc. Initial seeks in particular *should* work and hopefully my explanation helps to understand how.

Please be extremely careful with any code touching the seek/flush code paths.

> I agree, but currently, in MSE implementation, we send the seek event directly to the WebKitMediaSrc element.
> Consenquetly the elements don't receive the seek event and receive directly the FLUSH event, thus they can't store the seek event when they are not ready.

I will clarify some things regarding the above comment:

* WebKitMediaSrc does ***not*** send seek events downstream. Seeks are handled by the WebKitMediaSrc element, not the decoders or sinks. In regular playback, seeks are usually handled by the demuxer. In general, the element responsible of seeks is the one that has some access to the sample table.
  * It's common in GStreamer players that the sink element responsible for the video interface sends `seek` events upstream, but the intention here is to find the element responsible for handling seeks. Decoders should propagate but ignore `seek` events.
  * This is not the case in MediaPlayerPrivateGStreamerMSE, as we send the seek directly to the source element. The reason for this is that, at least as of the date the code was written,  playbin doesn't propagate initial seeks, therefore we needed to send them ourselves.
* Unlike seek events, flush events `do` propagate downstream. When a seek is done on WebKitMediaSrc at any point after the first sample has been popped from the TrackQueue, a FLUSH_START is sent from the main thread to all downstream elements, followed by a FLUSH_STOP as well. This is generally how a GStreamer element is meant to handle a seek event. After the FLUSH_STOP a `segment` event is sent from the streaming thread containing the new playback position.

**Comment by ntrrgc** [comment]
> For reference, this relates to the following downstream MR/issue:
> 
>   * [[wpe-2.28] Momentary freeze was observed on DAZN app assets WebPlatformForEmbedded/WPEWebKit#867](https://github.com/WebPlatformForEmbedded/WPEWebKit/issues/867)
> 
>   * [Fix seeking on DAZN & AppleTV+ apps WebPlatformForEmbedded/WPEWebKit#868](https://github.com/WebPlatformForEmbedded/WPEWebKit/pull/868)

These issues relate to wpe-2.28, which ships an older version of WebKitMediaSrc based on `appsrc`.

Current upstream has a completely reworked WebKitMediaSrc element that is no longer based on `appsrc`.

https://github.com/WebPlatformForEmbedded/WPEWebKit/blob/wpe-2.28/Source/WebCore/platform/graphics/gstreamer/mse/WebKitMediaSourceGStreamer.h
https://github.com/WebKit/WebKit/blob/main/Source/WebCore/platform/graphics/gstreamer/mse/WebKitMediaSourceGStreamer.h

Assuming the purpose of this PR was to upstream a fix for wpe-2.28 WebKitMediaSrc, that should be not necessary. Please reopen otherwise.

**Comment by ybandou** [comment]
@ntrrgc Thank you very much for your detailed explanation, it is vrey clear.

At the moment I don't have much time to check again, I will do it asap.
But I am sure we had an issue with DAZN app on WPE2.32 and I fixed it with this PR.

**Comment by ntrrgc** [comment]
> But I am sure we had an issue with DAZN app on WPE2.32 and I fixed it with this PR.

What was the issue on more technical terms?

**Pull request closed without merging**